designs/sam-config.md

```diff
@@ -0,0 +1,413 @@
+..
+..
+..
+```
+
+Why samconfig under `.aws-sam`
+---------------------------------
+
+> The below two don't answer the above question. Instead, they give technical implementation answers to questions not asked like "How is the .aws-sam directory made?" and "Are any files in .aws-sam git ignored?". The ideas surfaced in the previous PR's discussion were lost. For example, in the PR threads it was for a short time in the root of the project but then put back into .aws-sam. I recommend clearly stating here an answer to the above question in terms of goals, not technical implementations.
+
+The `.aws-sam` directory within the project directory is created with normal 755 permissions as default without any special permisions. `sam build` only creates a build directory within `.aws-sam` as `.aws-sam/build`. This directory is erased and re-built on every build. but top level directory is left unaffected.
+
+The `.gitignore` specified in the init apps also only have `.aws-sam/build` ignored and not anything else.
+
+
+Config file versioning
+-----------------------
+
+The configuration file: `samconfig.toml` will come with a top level version key that specifies the grammar version of the configuration file. This version can then be used to determine if a given configuration file works with a version of SAM CLI. A given version of SAM CLI may support one or more configuration file grammar versions.
+
+> The above clarifies that the version is not the version of the file itself (like an ever increasing commit version), instead it is the version of the grammar/DTD/spec of the file.
+
+It also paves the forward when major changes need to be made to the configuration file and add a version bump to the config file version
+
+```
+version = 0.1
+```
+
+> What is behavior of SAM CLI for different scenarios, e.g.:
+> 1. SAM CLI given file with version number that this SAM CLI doesn't support
+> 2. SAM CLI parses file but discovers that file doesn't comply with spec of that version
+> 3. SAM CLI tries to merge multiple configuration files together of differing spec versions
+
+Overrides
+----------
+
+The default location of a .aws-sam/samconfig can be replaced by overriding an environment variable called `SAM_CLI_CONFIG`
+
+`
+export SAM_CLI_CONFIG=~/Users/username/mysamconfig.toml
+`
+
+Users can pass an environment `--env` for the section that will be scanned within the configuration file to pass parameters through.
+
+> I hesitate in using `--env` as the param that controls what prefix of TOML tables/sections is parsed/loaded. Using `--env` strongly associates a specific usage scenario (environment choice) to a feature that is much broader in potential usage. Also, `--env` doesn't to me clarify what it controls/changes. At first read, I think it controls environment variables. Perhaps an alternative like `--config-prefix` which is no longer limited to a specific usage scenario while more clearly describing what it controls.
```

Comment:
Yes, this is the reason why the `--env` is hidden and ignored right now. To be able to support a better naming nomenclature.